grudge/discretization.py

```diff
@@ -636,17 +563,166 @@ def normal(self, dd):
 
     # }}}
 
+# }}}
+
+
+# {{{ Discretization construction routines
+
+def make_discretization_collection(
+        array_context: ArrayContext,
+        discr_context,
+        order=None,
+        discr_tag_to_group_factory=None,
+        mpi_communicator=None) -> DiscretizationCollection:
+    """Construct a discretization collection given an array context *array_context*
+    and user-provided discretization parameters.
+
+    :arg discr_context: A :class:`~meshmode.mesh.Mesh` object to define
+        the discretization collection over.
```

Comment:
Not sure I love the name. In pytential, @alexfikl used "places": https://documen.tician.de/pytential/symbolic.html in a similar role.

Reply:
In particular, the [GeometryCollection](https://documen.tician.de/pytential/symbolic.html#pytential.GeometryCollection) class in there. (The name was used in the code before introducing the GeometryCollection, so I just named it like that to stay consistent)

Reply:
To be honest, I'm starting to like this placeholder business less and less the more I think about it. We need to get this right. `make_discretization_collection` is intended to be the primary user-facing way to create the `DiscretizationCollection` And I can already imagine having an ambiguous placeholder argument causing confusion later on down the line.

I am not a fan, and I think we should avoid them if we can. I would rather just rename `discr_context` back to `mesh`, document it as such, and add an optional kwarg for discretization-related metadata that can be leveraged in the future. Thoughts @inducer @alexfikl ?

Reply:
@majosm This is probably a relevant discussion.

Reply:
> I'm starting to like this placeholder business less and less the more I think about it. 

Not sure I understand what you're objecting to. Having symbolic names for parts of the domain? Something else? Sorry if I'm being dense.

Reply:
Sorry if I wasn't very clear. I was referring to using placeholder arguments in `make_discretization_collection` like `places` or `discr_context` (determined by position alone) which can be pretty much anything (mesh, discretizations, dicts of discretizations). I don't feel comfortable doing that.

Reply:
I would rather try to use something along the lines of `**kwargs`, which is a bit more controlled from my perspective.

Reply:
Personally, I feel like I've had OK experiences with those types of arguments, if used judiciously. Examples:

- `loopy` functions will often take objects or string representations that parse to them. (e.g. domains, statements)
- `pyopencl.Array`'s constructor can take a context or an array as first argument.

The type of an object can be unambiguously determined, so I don't see a big risk in taking e.g. `mesh` as a shorthand for `{DEFAULT_VOLUME: mesh}`.

On the other hand, I find `**kwargs` parsing kind of unattractive:
- Forget to check for extra arguments
- Assuming you mean we would use `mesh=` or `geometries=` as alternatives to each other, we'd need to add logic to reject calls that have both.
- `**kwargs` have to be the last argument. If that makes the interface kwargs-only, then all calls become kind of wordy, which isn't amazing for high-profile API real estate like the effective discretization collection constructor.